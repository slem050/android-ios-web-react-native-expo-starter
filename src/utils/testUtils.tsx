import { PortalProvider } from '@gorhom/portal'
import { NavigationContainer } from '@react-navigation/native'
import { render, RenderAPI } from '@testing-library/react-native'
import { PropsWithChildren, ReactElement } from 'react'
import { I18nextProvider } from 'react-i18next'

import i18n from '~i18n/i18nForTests'
import { AuthProvider } from '~providers'
import { ColorSchemeProvider } from '~providers/ColorSchemeProvider'

type RenderOptions = Parameters<typeof render>[1]

const ProvidersWrapper: React.FC<PropsWithChildren> = ({ children }) => (
  <PortalProvider>
    <AuthProvider>
      <NavigationContainer>
        <ColorSchemeProvider>
          <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
        </ColorSchemeProvider>
      </NavigationContainer>
    </AuthProvider>
  </PortalProvider>
)

const customRender = (ui: ReactElement, options?: RenderOptions): RenderAPI => {
  return render(ui, { wrapper: ProvidersWrapper, ...options })
}

export * from '@testing-library/react-native'
export { customRender as render }
