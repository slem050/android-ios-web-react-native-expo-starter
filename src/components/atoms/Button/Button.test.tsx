import { StyleSheet } from 'react-native'

import { Button } from '.'

import { _appTheme } from '~constants'
import { cleanup, render, fireEvent, act } from '~utils/testUtils'

afterEach(cleanup)
describe('Button', () => {
  it('renders correctly', () => {
    const { toJSON } = render(<Button title="Button" />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('renders correctly Button.Primary', () => {
    const { getByTestId } = render(<Button.Primary title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      alignItems: 'center',
      backgroundColor: _appTheme.colors.primary,
      borderColor: 'transparent',
      borderRadius: 4,
      borderWidth: undefined,
      flexDirection: 'row',
      justifyContent: 'center',
      minWidth: 128,
      paddingHorizontal: 24,
      paddingVertical: 8,
    })
  })

  it('renders correctly Button.Secondary', () => {
    const { getByTestId } = render(<Button.Secondary title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      alignItems: 'center',
      backgroundColor: _appTheme.colors.secondary,
      borderColor: 'transparent',
      borderRadius: 4,
      borderWidth: undefined,
      flexDirection: 'row',
      justifyContent: 'center',
      minWidth: 128,
      paddingHorizontal: 24,
      paddingVertical: 8,
    })
  })

  it('renders correctly Button.Outline', () => {
    const { getByTestId } = render(<Button.Outline title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      alignItems: 'center',
      backgroundColor: 'transparent',
      borderColor: _appTheme.colors.primary,
      borderRadius: 4,
      borderWidth: StyleSheet.hairlineWidth,
      flexDirection: 'row',
      justifyContent: 'center',
      minWidth: 128,
      paddingHorizontal: 24,
      paddingVertical: 8,
    })
  })

  it('renders correctly Button.Ghost', () => {
    const { getByTestId } = render(<Button.Ghost title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      alignItems: 'center',
      backgroundColor: 'transparent',
      borderColor: 'transparent',
      borderRadius: 4,
      borderWidth: undefined,
      flexDirection: 'row',
      justifyContent: 'center',
      minWidth: 128,
      paddingHorizontal: 24,
      paddingVertical: 8,
    })
  })

  it('renders correctly Button.Link', () => {
    const { getByTestId } = render(<Button.Link title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      alignItems: 'center',
      backgroundColor: 'transparent',
      borderColor: 'transparent',
      borderRadius: 4,
      borderWidth: undefined,
      flexDirection: 'row',
      justifyContent: 'center',
      minWidth: 128,
      paddingHorizontal: 24,
      paddingVertical: 8,
    })
  })

  it('should call onPress', async () => {
    const onPress = jest.fn()
    const { getByTestId, update } = render(<Button title="Button" onPress={onPress} />)
    const button = getByTestId('baseButton')
    fireEvent.press(button)
    await act(async () => {
      update(<Button />)
    })
    expect(onPress).toHaveBeenCalledTimes(1)
  })

  it('should not call onPress when disabled', async () => {
    const onPress = jest.fn()
    const { getByTestId, update } = render(<Button title="Button" disabled onPress={onPress} />)
    const button = getByTestId('baseButton')
    fireEvent.press(button)
    await act(async () => {
      update(<Button />)
    })
    expect(onPress).toHaveBeenCalledTimes(0)
  })
})
