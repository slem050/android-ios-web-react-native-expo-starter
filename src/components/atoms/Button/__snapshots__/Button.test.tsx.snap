// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#3b82f6",
      "borderColor": "transparent",
      "borderRadius": 4,
      "borderWidth": undefined,
      "flexDirection": "row",
      "justifyContent": "center",
      "minWidth": 128,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
  testID="baseButton"
>
  <Text
    allowFontScaling={false}
    style={
      {
        "color": "#ffffff",
        "fontFamily": "Lato-Regular",
        "fontSize": 16,
        "fontStyle": "normal",
        "fontWeight": "400",
        "lineHeight": 21,
        "textAlign": "center",
        "textTransform": "none",
      }
    }
    testID="baseText"
    textAlign="center"
  >
    Button
  </Text>
</View>
`;
